File: Algorithm/C++/1000+/1001. Grid Illumination.cc
/*
 * Created on Wed Feb 27 2019 20:53:54
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// record the number of lights at each row, column, left diagonal and right diagonal
// use hashmap instead of list to avoid memory limit exceeded
#include <vector>
#include <unordered_map>
#include <unordered_set>

using std::vector;
using std::unordered_map;

class Solution {
 public:
  vector<int> gridIllumination(int N, vector<vector<int>>& lamps, vector<vector<int>>& queries) {
    unordered_map<int, int> row, col, left_diagonal, right_diagonal;
    unordered_map<int, std::unordered_set<int>> source;
    for (auto e : lamps) {
      source[e[0]].insert(e[1]);
      row[e[0]]++;
      col[e[1]]++;
      left_diagonal[e[0]+e[1]]++;
      right_diagonal[e[0]-e[1]]++;
    }
    vector<int> result;
    for (auto q : queries) {
      if (row[q[0]] > 0 || col[q[1]] > 0 || left_diagonal[q[0]+q[1]] > 0 || right_diagonal[q[0]-q[1]] > 0) {
        result.push_back(1);
      } else {
        result.push_back(0);
      }
      for (int i = q[0]-1; i <= q[0]+1; ++i) {
        for (int j = q[1]-1; j <= q[1]+1; ++j) {
          if (source.count(i) > 0 && source[i].count(j) > 0) {
            row[i]--;
            col[j]--;
            left_diagonal[i+j]--;
            right_diagonal[i-j]--;            
          }
        }
      }
    }
    return result;
  }
};

File: Algorithm/Python/1000+/1001. Grid Illumination.py
# -*- coding: utf-8 -*-
# Created on Wed Feb 27 2019 8:40:28
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# record the number of lights at each row, column, left diagonal and right diagonal
# use hashmap instead of list to avoid memory limit exceeded

from collections import defaultdict

class Solution(object):
    def gridIllumination(self, N, lamps, queries):
        """
        :type N: int
        :type lamps: List[List[int]]
        :type queries: List[List[int]]
        :rtype: List[int]
        """
        row, col = defaultdict(int), defaultdict(int)
        left_diagonal, right_diagonal = defaultdict(int), defaultdict(int)
        source = defaultdict(set)
        for i, j in lamps:
            source[i].add(j)
            row[i] += 1
            col[j] += 1
            left_diagonal[i+j] += 1
            right_diagonal[i-j] += 1

        result = []
        for i, j in queries:
            if row[i] > 0 or col[j] > 0 or left_diagonal[i+j] > 0 or right_diagonal[i-j] > 0:
                result.append(1)
            else:
                result.append(0)
            for r in [i-1, i+2]:
                for c in [j-1, j+2]:
                    if c in source[r]:
                        row[r] -= 1
                        col[c] -= 1
                        left_diagonal[r+c] -= 1
                        right_diagonal[r-c] -= 1
        return result
